src/simple.ts

```diff
@@ -3,40 +3,54 @@ import { BytesLike, HexString, toHex } from './bytes';
 import { MultiProof, processProof, processMultiProof } from './core';
 import { MerkleTreeData, MerkleTreeImpl } from './merkletree';
 import { MerkleTreeOptions } from './options';
+import { NodeHash } from './hashes';
 import { validateArgument } from './utils/errors';
 
 export interface SimpleMerkleTreeData extends MerkleTreeData<HexString> {
   format: 'simple-v1';
+  hash?: 'custom';
+}
+
+export interface SimpleMerkleTreeOptions extends MerkleTreeOptions {
+  nodeHash?: NodeHash;
 }
 
 export function formatLeaf(value: BytesLike): HexString {
   return defaultAbiCoder.encode(['bytes32'], [value]);
 }
 
 export class SimpleMerkleTree extends MerkleTreeImpl<BytesLike> {
-  static of(values: BytesLike[], options: MerkleTreeOptions = {}): SimpleMerkleTree {
-    const [tree, indexedValues] = MerkleTreeImpl.prepare(values, options, formatLeaf);
-    return new SimpleMerkleTree(tree, indexedValues, formatLeaf);
+  static of(values: BytesLike[], options: SimpleMerkleTreeOptions = {}): SimpleMerkleTree {
+    const [tree, indexedValues] = MerkleTreeImpl.prepare(values, options, formatLeaf, options.nodeHash);
+    return new SimpleMerkleTree(tree, indexedValues, formatLeaf, options.nodeHash);
   }
 
-  static load(data: SimpleMerkleTreeData): SimpleMerkleTree {
+  static load(data: SimpleMerkleTreeData, nodeHash?: NodeHash): SimpleMerkleTree {
     validateArgument(data.format === 'simple-v1', `Unknown format '${data.format}'`);
-    return new SimpleMerkleTree(data.tree, data.values, formatLeaf);
+    validateArgument(
+      (nodeHash == undefined) !== (data.hash == 'custom'),
+      nodeHash ? 'Data does not expect a custom node hashing function' : 'Data expects a custom node hashing function',
+    );
+
+    const tree = new SimpleMerkleTree(data.tree, data.values, formatLeaf, nodeHash);
+    tree.validate();
+    return tree;
```

Comment:
@ernestognw @frangio Any opinion on validating the tree here, has opposed to expecting the user calls `validate()` after loading

Reply:
I'd rather do it in `MerkleTree`'s constructor. We're already validating arguments there and the validation makes sense regardless of the node/leaf hashing functions

Reply:
Solving this because I see no reason not to add the check to construction. Feel free to reopen if anyone has a different opinion

Reply:
I considered checking in the constructor but:
- the constructor is protected, so that is not an entry point
- entry points are `.of(...)` and `.load(...)`
  - `.of(...)` builds the data correctly. We don't really need to re-validate (tests do that, but no need to do it in production)
  - `.load(...)` takes data taht may be incorrect, so it makes sens to check here.
  
That is why I put the check in `load` and not in the constructor.

Reply:
Right, these are good points, but:

- I think it's not an issue if we validate in each construction, it's just making it a bit less performant but that's fine in Javascript imo
- If we set the validation in the `of` function, we'll have to do it for each tree, I thought it make sense to just keep the validation at construction
- We don't recommend it, but people can extend the trees and write their static functions. Validating at construction would protect users against honest mistakes when making a customization.

Reply:
I am not sure how fast/slow construction of the tree (using `.of()`) is for "large" sets of leaves.

Reply:
Some people may argue its valuable to have a `load` that does not validate when you trust the input and want to go fast

Reply:
Fair, I wonder if people would use a Javascript library if they want to go fast anyway.

Reply:
Is there any good alternative, that does the same as our library, in another langage?

Also, building the tree might be slow, but then you dump it to json, and any React/Vue/... frontend can load the dump using our lib so that the user generate the proof locally in their browser for claiming some airdrop.

Reply:
Is to provide a "fullstack" solution. Fullstack includes frontend ... and frontend very often means JS/TS.

Reply:
> Is there any good alternative, that does the same as our library, in another langage?

This is also a good point. I mostly think speed is not a concern when doing fullstack development, and we're balancing speed vs (arguably) security in this case.

Not a strong opinion, I'm in the middle now. Let's discuss it in the weekly sync.

Reply:
I'd love @frangio opinion also. He may have been the one that choose to not include `.validate()` in `.load(...)` in the first place.

Reply:
Also valid point. 

Btw, I'm realizing both validating in construction and `load` are strictly breaking because they're changing the semantics of the function. That's a good reason to keep the validate in `load` so we minimize the impact.

If there was a reason why @frangio pushed for not validating on `.load`, let's wait for his input

Reply:
> I wonder if people would use a Javascript library if they want to go fast anyway.

Yes!! JavaScript itself _is_ fast. As an ecosystem it is only slow because libraries are slow, so we should do our part here and design/implement with performance in mind.

+1 to calling `validate` in `load` automatically.

Reply:
Right, I also feel Javascript is fast enough but bloated by libraries. Just wasn't sure it was our priority in this case.

Got it. I'll revert the last commit and then we're good to go.